### ISSUE TITLE
Socket for inbound Artery connections are not closed during coordinated shutdown

### ISSUE THREAD

**Issue by baltiyskiy** [akka/akka#30188]
We are observing a problem in our Akka cluster deployed in Kubernetes, where a pod, when restarted, cannot join cluster for ~15 minutes. Akka version is 2.5.23. We have been able to gather the following observations:

1. The joining node cannot join cluster because it tries to send InitJoin, but it never reaches the seed node.
2. We are currently deploying apps as statefulsets, so after a restart the pod keeps the address. 
3. The joining node attempts to establish the outbound control connection to the seed node, but fails repeatedly. The situation resolves when it finally succeeds in establishing the outbound control connection.
4. The joining node cannot establish the outbound control connection to the seed node because it establishes a TCP connection to the seed node and sends a HandshakeReq, but it never receives a HandshakeRsp. It should come through the inbound control connection, which the seed node should establish to the joining node.
5. We are running akka nodes as statefulsets, so the joining node keeps the address after restart.
5. The seed node does not establish the inbound control connection to the new incarnation of the joining node, because when the seed receives HandshakeReq from the joining node, the seed finds the existing association by the joining node's address, and then the seed uses the existing outbound control connection to the joining node, and sends HandshakeRsp on that connection.
6. (!) Here's the first problem --- the seed node should have dropped the outbound control connection to the joining node the moment it registers the new incarnation. But it doesn't do that. 
7. Why is the outbound control connection from the seed node to the joining node still around? Because when the old incarnation of the joining node was shutting down, it didn't close the socket! Then, thanks to Docker's virtual network interfaces, the socket still exists after the container is shut down. (See also this article: https://blog.box.com/container-networking-mystery-missing-rsts). From the seed node's viewpoint, the connection to the old incarnation is still ESTABLISHED, and it keeps retransmitting messages until it runs out of retransmit attempts --- it happens after ~15 minutes --- then it finally drops the connection, and finally attempts to establish a new connection to the joining node.
8. `tcpdump` shows that no FIN packets are sent by the node that goes down, nor RST packets. `strace` shows that the application (the joining node) does not call `shutdown` on the socket of the inbound control connection.
9. (!) Closing the inbound connections is actually not a part of the coordinated shutdown. It only happens because the TcpIncomingConnection actor is stopped when the actor system is stopped. In its `postStop()` method, it first calls `prepareAbort()` where it sets `SO_LINGER` to `true` with zero timeout (good), but then the actual `Socket.close()` is called asynchronously - it schedules a task on the executor (`akka.io.SelectionHandler.ChannelRegistryImpl#cancelKeyAndClose`).

My hunch is that by the time `akka.io.SelectionHandler.ChannelRegistryImpl#cancelKeyAndClose` is called, the executor is stopped.

**Comment by patriknw** [comment]
Thanks for the detailed report. As first step I'd recommend that you try with latest 2.5.32 and eventually update to Akka 2.6.x.

**Comment by baltiyskiy** [comment]
I've tried 2.5.32 before investigating, it didn't help.

**Comment by baltiyskiy** [comment]
I've noticed that the task that is scheduled inside `akka.io.SelectionHandler.ChannelRegistryImpl#cancelKeyAndClose` **is** executed, but the `shutdown()` syscall is still not issued, so the problem is not there.

I've added a log message after `key.channel().close()` (the call to `SocketChannel.close()`) and also logged `SelectionKey`. I see that log message, so the socket is indeed in the closed state. The state of the key before closing the channel: 
````
channel=java.nio.channels.SocketChannel[connected local=/10.32.255.39:2551 remote=/10.32.237.14:60244], selector=sun.nio.ch.EPollSelectorImpl@3326f03c, interestOps=1, readyOps=1
````

So maybe the problem is that `selector.select()` isn't called after that, because the whole system is going down. Is it plausible?

**Comment by jrudolph** [comment]
I'd say orphaned TCP connections are a thing and we should find ways to detect a broken connection that doesn't rely on RST or FIN packets arriving.

Here are a few options we should check:

 * The simplest approach would probably be to enable TCP keep-alive on remoting connections with timeout values that make sense given the expected pause times and needed reactivity. It seems the actual timeouts can only be configured starting from JDK 11 using `ExtendedSocketOptions.TCP_KEEP*` settings.

   The good thing about the TCP-stack keep-alive support is that it should be pretty transparent when it is enabled (connection just fails like in other cases of connection errors).
 * Are there application-level timeouts we could use to tear down that existing sending connection on the seed node earlier than only when the TCP connection fails after 15 minutes?
 * We should make sure the pair of connections is actually in a good state. Maybe be could avoid reusing an outgoing connection for another `HandshakeRsp`? I guess if both components of the connection pair are cut in a reasonable time frame (e.g. using keep-alives), it might not matter.

**Comment by jrudolph** [comment]
Two points I didn't address:

 * I didn't comment on the docker part. Not sure if it's really more common with virtual networks than with physical ones. Any IP that is still routeable but where no one is still answering might black hole packets (like mobile devices, unresponsive machines, etc), so the behavior doesn't seem a particular problem with docker (though, docker / k8s networking infrastructure probably could add rules to response with ICMP packets if a managed IP address is currently uninhabited, though not sure if it's economic to do so or if it would require to enumerate all inhabited or uninhabited addresses of a virtual network)
 * whether or not `SelectionHandler` may or may not be a problem. I agree it would be good to make sure that during GracefulShutdown all connections are closed properly. That's not sufficient, though, as such a shutdown process is often governed by all kinds of timeouts and a complete graceful shutdown cannot be guaranteed in all cases.

**Comment by baltiyskiy** [comment]
Hi Johannes,

thank you for weighing in! I would expect that once the in-cluster node learns about a new incarnation, it should drop all existing connections to it and re-create them. What would you say?

**Comment by jrudolph** [comment]
> thank you for weighing in! I would expect that once the in-cluster node learns about a new incarnation, it should drop all existing connections to it and re-create them. What would you say?

Yes, I think it should be done in `Association.completeHandshake`, do you agree @patriknw? I guess, we implicitly assume that the outbound has or will be closed automatically at some point, but might make sense to avoid using the same connection for different UniqueAddresses.

**Comment by baltiyskiy** [comment]
Just for your information: we've finally managed to find the root cause of the broken TCP connection.
In our case it wasn't the lack of connection shutdown. I was wrong in looking for `shutdown()` call; in fact, the way Akka uses Java sockets, we should expect `close()` to be called on the socket FD --- and it _is_ called. 
In our case, it was a bug in Calico: https://github.com/projectcalico/calico/issues/4518. Basically, when a pod is deleted, Calico would immediately (even before `preStop` hook is received by the application!) tear down the virtual network interface, marks the connection as invalid, and then prevents any packets to be either sent or received. 
The `close()` was in fact generating an RST packet, but it wasn't sent, so the other side never received it.

This also explains why I saw exiting-cluster phase time out: Akka node tries to notify other nodes in the cluster that the node is exiting and waits for the response; but since all connections are invalid, the packet was never sent.

However, it's still suspicious that Akka closes inbound connections as late as it does now --- even after the normal logging is torn down. I think it should be done during coordinated shutdown, not as a part of actor system shutdown (as it does now).

**Comment by jrudolph** [comment]
Thanks for digging down that hole. Good to know, that might come up in other situations.

> However, it's still suspicious that Akka closes inbound connections as late as it does now --- even after the normal logging is torn down. I think it should be done during coordinated shutdown, not as a part of actor system shutdown (as it does now).

But would that really matter in your case? Remoting is part of an actor system and has to be shut down at some point in the shutdown process. If access to a leaving node is just removed at some point, the remaining nodes might still be left in a half-finished communication state with the leaving node. So, IMO it doesn't make much sense trying to optimize the leaving node's behavior. Instead, we should think about how to improve behavior of the surviving nodes (e.g. see my previous comments above).

Another point: we recommend using deployments instead of statefulsets for Akka deployment. I assume that the problem is impossible or at least much more unlikely to happen with deployments because IP addresses will not (immediately) be reused. That's likely the reason that the problem has not come up for us.

**Comment by baltiyskiy** [comment]
As for deployments: yes, we're looking into moving to deployments. It didn't work when we started back in 2017, but possibly something has changed since then. 

> But would that really matter in your case? Remoting is part of an actor system and has to be shut down at some point in the shutdown process. If access to a leaving node is just removed at some point, the remaining nodes might still be left in a half-finished communication state with the leaving node. So, IMO it doesn't make much sense trying to optimize the leaving node's behavior. Instead, we should think about how to improve behavior of the surviving nodes (e.g. see my previous comments above).

Well, one might say that this is a part of a clean shutdown, just like the node announces that it is leaving the cluster during `exiting-cluster` phase.

**Comment by patriknw** [comment]
> As for deployments: yes, we're looking into moving to deployments

See https://developer.lightbend.com/docs/akka-platform-guide/deployment/index.html or https://doc.akka.io/docs/akka-management/current/kubernetes-deployment/index.html

The reason for shutting down remoting as late as possible is that actors are typically stopped after the coordinated shutdown phases when the actor hierarchy is stopped. They should have a possibility to send some some final messages when stopping (not guaranteed delivery of course).
